lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -291,6 +293,46 @@ void SPIRVRegularizeLLVMBase::lowerUMulWithOverflow(
   UMulIntrinsic->setCalledFunction(UMulFunc);
 }
 
+void SPIRVRegularizeLLVMBase::adaptStructTypes(StructType *ST) {
+  if (!ST->hasName())
+    return;
+  StringRef STName = ST->getName();
+  STName.consume_front("struct.");
+  StringRef MangledName = STName.substr(0, STName.find('.'));
+
+  // Demangle the name of a template struct and parse the template
+  // parameters which look like:
+  // <signed char, 2ul, 2ul, (spv::MatrixLayout)0, (spv::Scope)3>
+  // The result should look like SPIR-V friendly LLVM IR:
+  // %spirv.JointMatrixINTEL._char_2_2_0_3
+  if (MangledName.startswith("_ZTSN5__spv24__spirv_JointMatrixINTEL")) {
+    std::string DemangledName = llvm::demangle(MangledName.str());
+    StringRef Name(DemangledName);
+    Name = Name.slice(Name.find('<') + 1, Name.rfind('>'));
+    std::stringstream SPVName;
+    SPVName << kSPIRVTypeName::PrefixAndDelim
+            << kSPIRVTypeName::JointMatrixINTEL << kSPIRVTypeName::Delimiter;
+    auto P = Name.split(", ");
+    StringRef ElemType = P.first;
+    if (size_t Space = ElemType.rfind(' '))
+      ElemType = ElemType.substr(Space + 1);
+    SPVName << kSPIRVTypeName::PostfixDelim << ElemType.str();
+    P = P.second.split(", ");
+    StringRef Rows = P.first.take_while(llvm::isDigit);
+    SPVName << kSPIRVTypeName::PostfixDelim << Rows.str();
+    P = P.second.split(", ");
+    StringRef Cols = P.first.take_while(llvm::isDigit);
+    SPVName << kSPIRVTypeName::PostfixDelim << Cols.str();
+    P = P.second.split(", ");
+    StringRef Layout = P.first.substr(P.first.rfind(')') + 1);
+    SPVName << kSPIRVTypeName::PostfixDelim << Layout.str();
+    P = P.second.split(", ");
+    StringRef Scope = P.first.substr(P.first.rfind(')') + 1);
+    SPVName << kSPIRVTypeName::PostfixDelim << Scope.str();
+    ST->setName(SPVName.str());
```

Comment:
Wow, that is rough. Comments explaining which part of the demangling name we are trying to handle at each line would be helpful, especially if they all were referring to the same example you have atop of the enclosing if

Reply:
Added more comments in e549d01